[WebAssembly SIMD] Support vector comparisons on Intel

ddegazio opened the pull request:
#### c6145db418b325ac7054c3bddc544b59ac535163
<pre>
[WebAssembly SIMD] Support vector comparisons on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=248568">https://bugs.webkit.org/show_bug.cgi?id=248568</a>
rdar://103089559

Reviewed by Yusuke Suzuki.

Implements support for integer and floating point vector comparisons for the Intel
x86_64 macro assembler.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::compareFloatingPointVector):
(JSC::MacroAssemblerX86_64::compareIntegerVector):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpcmpeqq_rr):
(JSC::X86Assembler::vpcmpgtb_rr):
(JSC::X86Assembler::vpcmpgtw_rr):
(JSC::X86Assembler::vpcmpgtd_rr):
(JSC::X86Assembler::vpcmpgtq_rr):
(JSC::X86Assembler::vcmpps_rr):
(JSC::X86Assembler::vcmppd_rr):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDRelOp):

Canonical link: <a href="https://commits.webkit.org/257532@main">https://commits.webkit.org/257532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d21f9d01415bf7ef0a1d6d6b1385c8c972d6917

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108624 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168872 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8990 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85758 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91731 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106556 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104992 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89938 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2318 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85771 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2213 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28867 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5175 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42682 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88637 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3687 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19832 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -594,7 +594,48 @@ class AirIRGenerator {
             return { };
         }
         if (isValidForm(airOp, Arg::RelCond, Arg::SIMDInfo, Arg::Tmp, Arg::Tmp, Arg::Tmp)) {
+#if CPU(X86_64)
```
We usually use `if (isX86())` in WasmAirIRGenerator to keep both branches compiled / syntax-checked. (And it is constexpr).

ddegazio replied:
I'll switch it over, thanks for the review!

Constellation reviewed (approved):
r=me

ddegazio committed:
[WebAssembly SIMD] Support vector comparisons on Intel
https://bugs.webkit.org/show_bug.cgi?id=248568
rdar://103089559

Reviewed by Yusuke Suzuki.

Implements support for integer and floating point vector comparisons for the Intel
x86_64 macro assembler.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::compareFloatingPointVector):
(JSC::MacroAssemblerX86_64::compareIntegerVector):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpcmpeqq_rr):
(JSC::X86Assembler::vpcmpgtb_rr):
(JSC::X86Assembler::vpcmpgtw_rr):
(JSC::X86Assembler::vpcmpgtd_rr):
(JSC::X86Assembler::vpcmpgtq_rr):
(JSC::X86Assembler::vcmpps_rr):
(JSC::X86Assembler::vcmppd_rr):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDRelOp):

Canonical link: https://commits.webkit.org/257532@main
